src/core/thread/network_data_notifier.cpp

```diff
@@ -107,6 +117,13 @@ void Notifier::HandleNotifierEvents(Events aEvents)
         mNextDelay = 0;
     }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+    if (aEvents.ContainsAny(kEventThreadRoleChanged | kEventThreadNetdataChanged))
+    {
+        RequestRouterRoleIfEligible();
```

Comment:
If the leader does not support `kBorderRouterRequst` requests, we would keep sending requests in vain. Should we avoid that?

Reply:
In case a legacy Leader does not CoAP-ACK the request (because it does not recognize the kBorderRouterRequest status), there are also CoAP retransmissions sent.  A legacy Leader could also ACK the request but not send the response. Or it could ACK the request and send a response status of 'No Address Available'.  All of these cases may have a different timeout/retransmission behavior. Good to keep in mind.

In any case if the ADDR_SOL.req message has been ACK'ed by the Leader the BR but no Success response then the BR can at least assume the Leader's not willing to grant the request and could stop sending these -- until the Leader/Partition changes.

If the ADDR_SOL.req message is not ACK'ed by the Leader then the BR may also stop sending (like above) just for consistency and treat it as a denial of the request.   (Even though officially in CoAP this situation means that the CoAP request was never received by the Leader.)

Another solution could be to only retry such request again when the number of Routers in the Partition has changed. I.e. use that as a simple trigger to avoid sending requests in vain whenever some minor thing in the network data is changing.   New Leader/Partition is in any case a trigger to send the request again.

Reply:
Thanks guys. I agree we need to avoid retrying.

In the new push, I added a mechanism to limit sending such a request to once while local network data remains unchanged. This way if leader does not response we won't keep trying. Here is the comment about this:

```c++
    // We allow device to request router role upgrade using status
    // reason `kBorderRouterRequst` once while its local network data
    // remains unchanged. This ensures if the leader is running an
    // older version of Thread stack which does not support
    // `kBorderRouterRequst` reason, we do not keep trying (on no
    // response). The boolean `mDidRequestRouterRoleUpgrade` tracks
    // this. It is set to `false` when local network data get changed
    // or when partition ID gets changed (indicating a potential
    // leader change).
```

I think @EskoDijk described leader behavior nicely. 😃 
The OT code before https://github.com/openthread/openthread/pull/7578 would simply ignore the request (no response would be sent and reading the `Coap` code, I think no CoAP ack either).  But from sender side, the request will be timed out after some number of attempts (by `Coap` layer). Since we now limit this to one request, it should be ok to try once and time out.

Reply:
SGTM. Sending once after local network data change should work fine.